Source/DafnyLanguageServer.Test/Lookup/ReferencesTest.cs

```diff
@@ -0,0 +1,278 @@
+      var source = @"
+datatype Letter = ><A | B | C
+method M(l: Letter) {
+  print match l
+    case [>><A<] => 1
+    case B => 2
+    case C => 3
+  ;
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task DatatypeDestructor() {
+      var source = @"
+datatype Option = None | Some(><value: int)
+method M(o: Option) {
+  print match o
+    case None => 0
+    case Some(value) => value + o.[>><value<]
+  ;
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    // It seems that type declaration/usage information is not tracked
+    [Fact(Skip = "Not implemented")]
+    public async Task Type() {
+      var source = @"
+type ><T = int
+method M(t: [>><T<]) {}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    // It seems that newtype declaration/usage information is not tracked
+    [Fact(Skip = "Not implemented")]
+    public async Task Newtype() {
+      var source = @"
+newtype ><T = int
+method M(t: [>><T<]) {}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Trait() {
+      var source = @"
+trait ><T {}
+class C extends [>><T<] {}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    // It seems that class declaration/usage information is not tracked
+    [Fact(Skip = "Not implemented")]
+    public async Task ClassDeclaration() {
+      var source = @"
+class ><C {}
+method M(c: [>><C<]) {}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task ClassField() {
+      var source = @"
+class C {
+  var ><f: int
+  constructor() { [>><f<] := 0; }
+}
+method M(c: C) {
+  print c.[>><f<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task ClassMethod() {
+      var source = @"
+class C {
+  method ><CM() {}
+}
+method M(c: C) {
+  c.[>><CM<]();
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    // TODO - not sure why both primary assertions are failing
+    [Fact]
+    public async Task AcrossFiles() {
+      var sourceA = @"
+include ""B.dfy""
+module ><A {}
+".TrimStart();
+      var sourceB = @"
+include ""A.dfy""
+module B {
+  import [>><A<]
+}
+".TrimStart();
+
+      MarkupTestFile.GetPositionsAndRanges(
+        sourceA, out var cleanSourceA, out var positionsA, out var rangesA);
+      MarkupTestFile.GetPositionsAndRanges(
+        sourceB, out var cleanSourceB, out var positionsB, out var rangesB);
+      Assert.Single(positionsA);
+      Assert.Single(positionsB);
+      Assert.Single(rangesB);
+
+      var cwd = Directory.GetCurrentDirectory();
+      var pathA = Path.Combine(cwd, "Lookup/TestFiles/A.dfy");
+      var pathB = Path.Combine(cwd, "Lookup/TestFiles/B.dfy");
+
+      var documentItemA = CreateTestDocument(cleanSourceA, pathA);
+      await client.OpenDocumentAndWaitAsync(documentItemA, CancellationToken);
```

Comment:
Only `pathA` is visible to the language server at this point right? `pathB` is neither in memory nor on disc

Reply:
Thanks, I misunderstood that `CreateTestDocument` doesn't automatically make the contents available for the resolver/LS to read. I've created appropriate files in `TestFiles` to make the test work.

Reply:
As long as you keep the document open after using `client.OpenDocumentAndWaitAsync`, it should be visible to all projects, as if it is on disk, so that's an alternative.